### ISSUE TITLE
Remove remaining boost parts with the standard library

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#3095]
Hi,

There are still parts of boost library that are used inside mlpack, more specifically in here
- [x] `mlpack/src/mlpack/core/tree/cosine_tree/cosine_tree.hpp`  boost::heap
- [ ] `mlpack/src/mlpack/core/data/load_arff.hpp`        boost::tokenizer
- [x] `mlpack/tests/main_tests/emst_test.cpp`              boost::math::iround()
- [x] `mlpack/core/tree/cosine_tree/cosine_tree.cpp`   boost::math::quantile()
- [x] `mlpack/methods/kde/kde_rules_impl.hpp`          boost::math::normal normalDist
- [ ] `mlpack/methods/kde/kde_rules_impl.hpp`          boost::math::quantile()

New one 
- [ ]  `mlpack/methods/reinforcement_learning/q_networks/categorical_dqn.hpp`  boost::get()
- [ ]  `mlpack/methods/reinforcement_learning/q_networks/dueling_dqn.hpp`  boost::get()
- [ ]  `mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp`  boost::get()
- [ ] `mlpack/bindings/markdown/print_docs.cpp`  boost::replace_all_copy()

We would like to replace this one with the standard library if possible. If someone is interested in taking this issue feel free to do so.

**Comment by rcurtin** [comment]
I think `boost::heap` can easily be replaced with `std::priority_queue`, and `boost::tokenizer` can be replaced pretty easily with `std::stringstream` and `std::getline()` or similar ([example](https://www.geeksforgeeks.org/tokenizing-a-string-cpp/)).  So, I think it should be relatively straightforward to fix this. :+1:

**Comment by heisenbuug** [comment]
@shrit should I take this one? It was marked as `good first issue` so not sure and seems pretty straightforward as well.

**Comment by shrit** [comment]
Yeah please go ahead. I do not have time at all, and I have a couple of PRs that are open and need to be finished first

**Comment by shrit** [comment]
Just found more `boost::math::normal` function inside the `kde_rules_impl`.
This thing is so invasive as a virus. :rofl:

**Comment by heisenbuug** [comment]
Yea, even we need to replace `boost::math::normal`.

I think we can go with `std::normal_distribution`, what we need to figure out is `boost::math::quantile`.

**Comment by conradsnicta** [comment]
> we need to figure out is `boost::math::quantile`.

Maybe [arma::quantile()](http://arma.sourceforge.net/docs.html#quantile) would help here.  Available in Armadillo since version 9.900.  The implementation is based on "Definition 5" in:

* Rob J. Hyndman and Yanan Fan. Sample Quantiles in Statistical Packages. The American Statistician, Vol. 50, No. 4, pp. 361-365, 1996. http://doi.org/10.2307/2684934

**Comment by shrit** [comment]
Hmm, that is interesting, however, we need to upgrade Armadillo version in this case. Maybe we can, as we still did not release mlpack 4

**Comment by heisenbuug** [comment]
So I will not work on `quantile` for now.
@rcurtin can we upgrade Armadillo version?

**Comment by rcurtin** [comment]
The problem is that Armadillo 9.900 is not available in Ubuntu 20.04 or RHEL8 (but both do have Armadillo version 9).  That would make it a bit more difficult to package when we finally do release mlpack 4, but maybe we can lean on our RHEL and Debian maintainers here (the first of those is me...) to upgrade Armadillo for the mlpack 4 release.

I dunno, I'm not excited about bumping the minimum Armadillo to something not available in distributions that people might reasonably be using, but, I guess that we have the autodownloader now.  So, perhaps the benefit of being able to remove another piece of boost easily outweighs the inconvenience for some users of not having the requisite version of Armadillo available.

That's a lot of words to say "sure, go for it", but I figured I'd think out loud. :)

**Comment by conradsnicta** [comment]
I concur that using Armadillo > 9.800 may be an issue.  If all that's required is the `arma::quantile()` function, it shouldn't be too difficult adapting the code from Armadillo (or Boost) and writing a specific mlpack version of the function, say `mlpack_quantile()`. 

By adapting I do not mean backporting the function from Armadillo 9.900 to 9.800.  Backporting invariably causes too many problems down the line.  It's better to write a completely new standalone function.  For reference, the core of the functionality is here: https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.7.x/include/armadillo_bits/glue_quantile_meat.hpp#L23-84

This is all under the assumption that `arma::quantile()` is actually doing the same (or very similar) thing as `boost::math::quantile`.

**Comment by conradsnicta** [comment]
Followup from the previous comment. So it looks like `arma::quantile()` and `boost::math::quantile` are not doing the same thing. The former takes a dataset (eg. vector of values generated from a distribution function), while the latter takes the specification of a distribution such as `boost::math::normal dist(mu, sigma)`.

**Comment by heisenbuug** [comment]
Then I guess we need to adopt one of the implementations for mlpack.
`boost::math::quantile()` is being used in 3 files.

I will have a look at both the implementations since we are not using `boost::math::quantile` in many places
even if we decide to adapt arma's implementation(with some small additional changes), it should be fine I guess.

**Comment by heisenbuug** [comment]
@shrit I added some more places to the issue description, hope that's fine.

**Comment by rcurtin** [comment]
Thanks for checking into that @conradsnicta.

I took a quick look and refreshed my memory, and our use of `boost::math::quantile()` is actually to compute the inverse of the CDF of a Gaussian distribution:

```
  // Fit a normal distribution using the calculated statistics, and calculate a
  // lower bound on the magnitudes for the passed 'delta' parameter.
  boost::math::normal dist(mu, sigma);
  double lowerBound = boost::math::quantile(dist, delta);
```

The `KDERules` usage is the same idea.

Wikipedia lists the quantile function of a Gaussian distribution as:

![image](https://user-images.githubusercontent.com/1845039/144150916-c02f8a06-5f61-461d-a0af-210539229d7d.png)

so I imagine (but have not verified) that what we need is the same thing.  @heisenbuug want to see what the `boost::math::quantile()` implementation actually implements for a Gaussian distribution?  If it's the formula above, then actually it seems like that should be pretty easy to implement ourselves.  https://en.cppreference.com/w/cpp/numeric/math/erfc may be helpful here too.

**Comment by shrit** [comment]
Just updated the list, I have noticed a new `boost::get()` in the categorical_dqn header.

**Comment by shubham1206agra** [comment]
I just want to ask why you want to remove the boost library usage from the specified location?

As I am seeing high usage of boost library in ANN modules.

**Comment by eshaanagarwal** [comment]
hi ! Is this issue open for contribution? I would like to work on this.

**Comment by heisenbuug** [comment]
> I just want to ask why you want to remove the boost library usage from the specified location?
> 
> As I am seeing high usage of boost library in ANN modules.

@shubham1206agra  We have another branch in which we are working on that.

**Comment by heisenbuug** [comment]
> hi ! Is this issue open for contribution? I would like to work on this.

Yea, you can.
Check [this](https://github.com/mlpack/mlpack/pull/3098) PR
I am working on some, you can start with `mlpack/tests/main_tests/emst_test.cpp boost::math::iround()`

**Comment by shubham1206agra** [comment]
> Thanks for checking into that @conradsnicta.
> 
> I took a quick look and refreshed my memory, and our use of `boost::math::quantile()` is actually to compute the inverse of the CDF of a Gaussian distribution:
> 
> ```
>   // Fit a normal distribution using the calculated statistics, and calculate a
>   // lower bound on the magnitudes for the passed 'delta' parameter.
>   boost::math::normal dist(mu, sigma);
>   double lowerBound = boost::math::quantile(dist, delta);
> ```
> 
> The `KDERules` usage is the same idea.
> 
> Wikipedia lists the quantile function of a Gaussian distribution as:
> 
> ![image](https://user-images.githubusercontent.com/1845039/144150916-c02f8a06-5f61-461d-a0af-210539229d7d.png)
> 
> so I imagine (but have not verified) that what we need is the same thing. @heisenbuug want to see what the `boost::math::quantile()` implementation actually implements for a Gaussian distribution? If it's the formula above, then actually it seems like that should be pretty easy to implement ourselves. https://en.cppreference.com/w/cpp/numeric/math/erfc may be helpful here too.

@rcurtin I think what we are looking for here is an approximation for probit function. Right?
If that's the case, I will implement the custom function for this. 
@heisenbuug Where should I put the function definition (in which file)?

**Comment by dnabanita7** [comment]
I guess this issue has been taken up by 2 people already.  Should I get started with some too?

**Comment by pree-T** [comment]
Is this issue still open?

**Comment by shrit** [comment]
it depends on what part you would like to handle

**Comment by Shashank-Shet** [comment]
Hello, I am a first-time contributor. Can I volunteer for this as well? If there are tasks remaining that is.

**Comment by heisenbuug** [comment]
Start with some part, make sure no one has commented that they are working on this one.
Check out related PR, this is for everyone who wants to contribute to this, there has been already some discussion done
on some parts, so make sure to keep those points in mind.



> > Thanks for checking into that @conradsnicta.
> > I took a quick look and refreshed my memory, and our use of `boost::math::quantile()` is actually to compute the inverse of the CDF of a Gaussian distribution:
> > ```
> >   // Fit a normal distribution using the calculated statistics, and calculate a
> >   // lower bound on the magnitudes for the passed 'delta' parameter.
> >   boost::math::normal dist(mu, sigma);
> >   double lowerBound = boost::math::quantile(dist, delta);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The `KDERules` usage is the same idea.
> > Wikipedia lists the quantile function of a Gaussian distribution as:
> > ![image](https://user-images.githubusercontent.com/1845039/144150916-c02f8a06-5f61-461d-a0af-210539229d7d.png)
> > so I imagine (but have not verified) that what we need is the same thing. @heisenbuug want to see what the `boost::math::quantile()` implementation actually implements for a Gaussian distribution? If it's the formula above, then actually it seems like that should be pretty easy to implement ourselves. https://en.cppreference.com/w/cpp/numeric/math/erfc may be helpful here too.
> 
> @rcurtin I think what we are looking for here is an approximation for probit function. Right? If that's the case, I will implement the custom function for this. @heisenbuug Where should I put the function definition (in which file)?

Hey sorry for the late reply. Been a bit busy.

Are you still working on this?
If yes, then you can create a new file for this function in `mlpack/core/math` I guess.
Also let me know if you are not working on this, I will get on with it then...

**Comment by Shashank-Shet** [comment]
So may I work on replacing the boost::get() parts of the code?

**Comment by heisenbuug** [comment]
@Shashank-Shet It's not taken by anybody so yes you can.

Just comment here how are you planning to implement or maybe you can ask your doubts on IRC channel as well.

**Comment by shubham1206agra** [comment]
> Start with some part, make sure no one has commented that they are working on this one. Check out related PR, this is for everyone who wants to contribute to this, there has been already some discussion done on some parts, so make sure to keep those points in mind.
> 
> > > Thanks for checking into that @conradsnicta.
> > > I took a quick look and refreshed my memory, and our use of `boost::math::quantile()` is actually to compute the inverse of the CDF of a Gaussian distribution:
> > > ```
> > >   // Fit a normal distribution using the calculated statistics, and calculate a
> > >   // lower bound on the magnitudes for the passed 'delta' parameter.
> > >   boost::math::normal dist(mu, sigma);
> > >   double lowerBound = boost::math::quantile(dist, delta);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The `KDERules` usage is the same idea.
> > > Wikipedia lists the quantile function of a Gaussian distribution as:
> > > ![image](https://user-images.githubusercontent.com/1845039/144150916-c02f8a06-5f61-461d-a0af-210539229d7d.png)
> > > so I imagine (but have not verified) that what we need is the same thing. @heisenbuug want to see what the `boost::math::quantile()` implementation actually implements for a Gaussian distribution? If it's the formula above, then actually it seems like that should be pretty easy to implement ourselves. https://en.cppreference.com/w/cpp/numeric/math/erfc may be helpful here too.
> > 
> > 
> > @rcurtin I think what we are looking for here is an approximation for probit function. Right? If that's the case, I will implement the custom function for this. @heisenbuug Where should I put the function definition (in which file)?
> 
> Hey sorry for the late reply. Been a bit busy.
> 
> Are you still working on this? If yes, then you can create a new file for this function in `mlpack/core/math` I guess. Also let me know if you are not working on this, I will get on with it then...

https://github.com/heisenbuug/mlpack/pull/2
I have already created a PR on your repo
I have 2 commit pending but I want first review it

**Comment by Ashutosh-codes** [comment]
Hello @shrit 
Can I contribute to boost::math::iround() part of the code?
This is my first contribution in open source. Any advice on how to  get started would  be very beneficial.
My domains of interest are Machine learning and deep learning and especially the mathematics and algorithms involved in it. I am also proficient in C++, python, pandas, numpy and matlab.
So, Can I start contributing to boost::math::iround() part of the  code?

**Comment by Shashank-Shet** [comment]
@heisenbuug  @shrit  I have been looking into finding a replacement for boost::get. I was looking into std::get as an alternative. However, the first issue with this is that it requires C++17 to compile. Do you have any directions in mind other than std::get for me to look into?

**Comment by Shashank-Shet** [comment]
Also, just another question, the code in question here is:
`      boost::get<NoisyLinear<>*>
          (network.Model()[noisyLayerIndex[i]])->ResetNoise();
`
Is there any scenario wherein `network.Model()` does not yield a vector?

**Comment by rcurtin** [comment]
I thought `std::get` was available from C++11.  In any case I wouldn't worry too much about the stuff in `src/mlpack/methods/ann/` because it is being revamped in #2779, and the work that is being done there is likely to remove that (and similar) `boost::get` calls anyway. :+1:

**Comment by rcurtin** [comment]
> > Start with some part, make sure no one has commented that they are working on this one. Check out related PR, this is for everyone who wants to contribute to this, there has been already some discussion done on some parts, so make sure to keep those points in mind.
> > > > Thanks for checking into that @conradsnicta.
> > > > I took a quick look and refreshed my memory, and our use of `boost::math::quantile()` is actually to compute the inverse of the CDF of a Gaussian distribution:
> > > > ```
> > > >   // Fit a normal distribution using the calculated statistics, and calculate a
> > > >   // lower bound on the magnitudes for the passed 'delta' parameter.
> > > >   boost::math::normal dist(mu, sigma);
> > > >   double lowerBound = boost::math::quantile(dist, delta);
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > The `KDERules` usage is the same idea.
> > > > Wikipedia lists the quantile function of a Gaussian distribution as:
> > > > ![image](https://user-images.githubusercontent.com/1845039/144150916-c02f8a06-5f61-461d-a0af-210539229d7d.png)
> > > > so I imagine (but have not verified) that what we need is the same thing. @heisenbuug want to see what the `boost::math::quantile()` implementation actually implements for a Gaussian distribution? If it's the formula above, then actually it seems like that should be pretty easy to implement ourselves. https://en.cppreference.com/w/cpp/numeric/math/erfc may be helpful here too.
> > > 
> > > 
> > > @rcurtin I think what we are looking for here is an approximation for probit function. Right? If that's the case, I will implement the custom function for this. @heisenbuug Where should I put the function definition (in which file)?
> > 
> > 
> > Hey sorry for the late reply. Been a bit busy.
> > Are you still working on this? If yes, then you can create a new file for this function in `mlpack/core/math` I guess. Also let me know if you are not working on this, I will get on with it then...
> 
> [heisenbuug#2](https://github.com/heisenbuug/mlpack/pull/2) I have already created a PR on your repo I have 2 commit pending but I want first review it

@shubham1206agra that PR looks good!  Do you think you could open it against the main mlpack repository instead?  I imagine it should be possible to consider the changes you made separate from the `remove-more-boost` branch (correct me if I'm wrong).

**Comment by a0n0k0i0t** [comment]
I would like to work on `boost::replace_all_copy()`. I'm not sure if there's any function in standard library that can replace it (correct me if I'm wrong) . I would like to implement a custom function using KMP algo. Can I work on it?

**Comment by shubham1206agra** [comment]
@a0n0k0i0t Actually I have already done this. Sorry for not giving heads up in this thread.

**Comment by a0n0k0i0t** [comment]
@shubham1206agra cools

**Comment by Ashutosh-codes** [comment]
I woould ike to work on boost::math::iround(). Can I go ahead on this issue?

**Comment by eshaanagarwal** [comment]
> I woould ike to work on boost::math::iround(). Can I go ahead on this issue?
I am working on this currently!

**Comment by Ashutosh-codes** [comment]
Hello @rcurtin I want to work on boost::tokenizer. Can I go ahead on this issue?

**Comment by kuries** [comment]
Hello everyone,
I wanted to know if anyone is currently working on the `boost::heap` removal or is it still up for grabs?

**Comment by heisenbuug** [comment]
> Hello everyone, I wanted to know if anyone is currently working on the `boost::heap` removal or is it still up for grabs?

Hi @kuries I was working on it, here is the [PR](https://github.com/mlpack/mlpack/pull/3098). It's not much but there are some comments which might help you get started.

I got busy with other things so you can take this up.

**Comment by kuries** [comment]
> > Hello everyone, I wanted to know if anyone is currently working on the `boost::heap` removal or is it still up for grabs?
> 
> Hi @kuries I was working on it, here is the [PR](https://github.com/mlpack/mlpack/pull/3098). It's not much but there are some comments which might help you get started.
> 
> I got busy with other things so you can take this up.

Sure, thank you 👍

**Comment by adithyathecoder** [comment]
I'm new to open source contribution it was shown good first issue can I know is this issue still open and if open how can I solve the issue like what exactly to do solve the issue

**Comment by Utkarsh2401** [comment]
Hi! I would like to start contributing to this issue, so if is there anything that no one is working on currently or which might require some help, please let me know.
Thank you.

**Comment by mohamedAdhamc** [comment]
I would like to take on this issue if possible I am still new to open source but i believe this would be a good first open source issue for me.

**Comment by shrit** [comment]
keep open

**Comment by conradsnicta** [comment]
Yet another good reason to remove Boost:  https://gitlab.com/conradsnicta/armadillo-code/-/issues/206

TLDR: looks like code rot is setting into Boost: https://en.wikipedia.org/wiki/Software_rot
